src/service/performance-impl.js

```diff
@@ -104,12 +105,23 @@ export class Performance {
     /** @private {string} */
     this.ampexp_ = '';
 
-    /** @private {number|null} */
-    this.makeBodyVisible_ = null;
-    /** @private {number|null} */
-    this.firstContentfulPaint_ = null;
-    /** @private {number|null} */
-    this.firstViewportReady_ = null;
+    const fcpDeferred_ = new Deferred();
```

Comment:
nit: instead of having `this.fcpPromise_` and `this.fcpResolve_`. You can store the `this.fcpDeferred_` and call `this.fcpDeferred_.promise` instead.

Reply:
same applies to other promises